Add Scores Radar view

alexandraabbas opened the pull request:
Adding a new view, Scores Radar. It visualizes metric values from a singles scorer (in a single eval) across models. Similar to Scores by Limit view, it has a pre-processing function which yields a dataframe that the view function can take as input.

**Considerations**
- I considered multiple ways (max, min-max, percentile scaling) to scale metric values to make comparing models straightforward. I settled with percentile scaling which converts each data point to its percentile position, which tells you what percentage of values in a dataset fall below the value. It gracefully handles negative values and prevents outliers which would squish the rest of the values to a narrow space.
- I updated `line()` and `text()` to be able to take `data=None` and `x, y, z, text=Sequence[Any]` to prevent needing to create a Data object for simple plotting needs.

**Future work**
- Support plotting across different scorers and evals.
- Extract generic radar chart core logic into a separate module (which would enable the above).
- Configure other suitable marks to take input arguments as sequences rather than a data object. I only configured `line()` and `text()` in this PR (since these are the ones used in the radar chart).

alexandraabbas committed:
Add scores radar view

alexandraabbas committed:
Correct typo

dragonstyle commented on src/inspect_viz/view/beta/_scores_radar.py:
```diff
@@ -0,0 +1,295 @@
+def scores_radar_df(
+    data: pd.DataFrame,
+    scorer: str,
+    metrics: list[str] | None | NotGiven = None,
```
Normally we include `NotGiven` when we need to differentiate between no metrics being passed and the user explicitly passing `None` (e.g. the user wants no metrics). In this case, I think None would mean use the default (there is no meaningful way to have this data frame without any metrics). I think you can remove `NotGiven`.

Similarly with invert, specifying `None` seems not necessary.

alexandraabbas replied:
Removed `NotGiven` from `metrics` and `invert`.

dragonstyle commented on src/inspect_viz/view/beta/_scores_radar.py:
```diff
@@ -0,0 +1,295 @@
+    invert: list[str] | None | NotGiven = None,
+    model_column_name: str = "model",
+    task_id_column_name: str = "task_id",
+    log_column_name: str = "log",
```
I think for these columns, you can just require columns with specific names:

```python
    # validate general columns
    required_columns = ["model", "task_id", score]
    for col in required_columns:
        if col not in df.columns:
            raise ValueError(f"Required column '{col}' not found in data.")
```

Similarly for `log` column name, I think you can just forward it along.

alexandraabbas replied:
Removed these input arguments.

dragonstyle commented on src/inspect_viz/view/beta/_scores_radar.py:
```diff
@@ -0,0 +1,295 @@
+        legend: Options for the legend. Pass None to disable the legend.
+        **attributes: Additional `PlotAttributes`.
+    """
+    if isinstance(data, pd.DataFrame):
```
Should validate the data frame we're receiving has the columns from the df function above.

alexandraabbas replied:
Validation added.

dragonstyle commented on src/inspect_viz/view/beta/_scores_radar.py:
```diff
@@ -0,0 +1,295 @@
+        columns_to_keep.append(log_column_name)
+
+    # handle multiple rows per model by taking first row
+    data_dedup = data.groupby(model_column_name)[columns_to_keep].first().reset_index()
```
Question: is this expected / should we instead throw here?

alexandraabbas replied:
It might happen that a user points to a log dir where there're multiple log files for the same eval-model pair. You're right that it makes sense to be explicit rather then taking the first row and ignoring the rest silently. I'll throw an exception here instead.

alexandraabbas committed:
Fix mypy numpy type error

dragonstyle commented on src/inspect_viz/view/beta/_scores_radar.py:
```diff
@@ -0,0 +1,295 @@
+
+def grid_circles_coordinates() -> list[dict[str, list[float]]]:
+    """Computes coordinates for grid circles to be used in a radar chart."""
+    radii = [0.2, 0.4, 0.6, 0.8, 1.0]
```
Should the number of range of circles coordinates be configurable?

alexandraabbas replied:
I don't see a use case for wanting to configure a different grid circles, it's fairly standard to have this setup (see [Matplotlib](https://matplotlib.org/stable/gallery/specialty_plots/radar_chart.html) for example). We can potentially ask for how many grid circles they want, but I would not expose this detail as an argument in scores_radar() view. I imagine, one could do the lower level plotting if they want to modify the base plot itself.

dragonstyle commented on src/inspect_viz/mark/_line.py:
```diff
@@ -43,6 +43,14 @@ def line(
+        data = args_to_data({"x": x, "y": y, "z": z})
+
+        # reassign parameters to column names for column_param
+        x = "x" if "x" in data.columns else None
```
I'm not fully following what these are doing... I'll ask you in realtime.

alexandraabbas replied:
I know these are not very nice. Let me think about a nicer solution. An option is to return a column mapping from `args_to_data`. When `data=None` and `x=[1, 2, 3]` and `y=[1, 2, 3]` we turn x and y into a new `Data` object called `data` here. We then need to reassign `x` and `y` to `"x"` and `"y"` (overwrite the sequences). If `z=None` (or any of `x`, `y` or `z`), it won't be added to `data` hence the `in data.columns else None` check.

dragonstyle commented on src/inspect_viz/mark/_line.py:
```diff
@@ -8,13 +8,13 @@
-    y: ChannelSpec | Param,
+    data: Data | None = None,
+    x: ChannelSpec | Param | None = None,
+    y: ChannelSpec | Param | None = None,
```
can x and y really be None?

alexandraabbas replied:
Well, in practice they cannot be None. However I'd need to change the function signature to allow data to be None but not x and y which would change the order of the input arguments. We'd need to update these across all usages of line. Shall we go with changing the function signature?

```
def line(
    x: ChannelSpec | Param,
    y: ChannelSpec | Param,
    data: Data | None = None,
```

dragonstyle reviewed (changes requested):
Here is a first pass (not having looked at docs)

alexandraabbas committed:
Address change requests

alexandraabbas committed:
Address change requests

alexandraabbas committed:
Fix test

alexandraabbas committed:
Address comments

dragonstyle commented on src/inspect_viz/mark/_text.py:
```diff
@@ -166,3 +175,25 @@ def text_y(
     return dict_to_camel(dict(styles)) if styles else {}
+
+
+def infer_text_mark_config_inputs(
```
What would you think about renaming this function to `resolve_inputs` and then making it so it is safe to always call (e.g. if data is not `None`, it just returns the current data and column values. This would let the calling code have:

```python
data, x, y, z, text = resolve_inputs(data, x, y, z, text)
```

with no conditional around it.

alexandraabbas replied:
Moved all related logic to `resolve_text_inputs`.

dragonstyle commented on src/inspect_viz/mark/_text.py:
```diff
@@ -146,6 +152,9 @@ def text_y(
         **options: Additional `MarkOptions`.
     """
+    if data is None:
+        data, x, y, z, text = infer_text_mark_config_inputs(data, x, y, z, text)  # type: ignore
```
This type ignore seems sketchy?

alexandraabbas replied:
I found a workaround, it's removed now.

dragonstyle commented on src/inspect_viz/mark/_line.py:
```diff
@@ -43,6 +43,17 @@ def line(
         tension: The tension option for bundle, cardinal and Catmull-Rom splines.
         **options: Additional `MarkOptions`.
     """
+    if data is None:
```
I think pulling this into a `resolve_inputs` function like you did for the text mark below would be an improvement

alexandraabbas replied:
Moved all related logic to `resolve_line_inputs`.

dragonstyle commented on src/inspect_viz/view/beta/_scores_radar.py:
```diff
@@ -0,0 +1,297 @@
+
+
+def scores_radar(
+    data: Data | pd.DataFrame,
```
If possible would be good to avoid accepting a DataFrame here unless we need it to support receiving a df. (All other functions just take Data)

alexandraabbas replied:
I think we discussed this previously if you remember. The function originally returned Data but I changed it to DataFrame to be able to save it as a parquet file, and to follow the convention of Scores by Limit view. I'd suggest keeping DataFrame otherwise we would need to save it as a parquet file in the function (as a side effect) before turning it into Data, unless I could save Data as a parquet file which I'm uncertain about. Afaik Data doesn't support saving to a file. _(A feature we might want to add?)_

dragonstyle replied:
Y, Data doesn't support saving from a file (which I think does make sense since Data includes things like selection which really don't seem meaningfully serializable). I'm thinking that `scores_radar_df` should return a DataFrame, but that `scores_radar` should accept only a Data (that way the user is required to make a Data which they can coordinate with other things like inputs should they choose).

dragonstyle commented on src/inspect_viz/view/beta/_scores_radar.py:
```diff
@@ -0,0 +1,297 @@
+    model: str = "model_display_name",
+    title: str | Title | None = None,
+    height: float | None = 700,
+    width: float | None = 700,
```
400x400 feels like a better default size, but when I manually set this, the legend gets very large and out of proportion (could be an issue with the legend sizing itself...). What do you think?

alexandraabbas replied:
Added adaptive scaling to `legend.ts` which prevents extremely small and large legends. The reason the previous approach didn't work for the radar plot was because how we calculate the scaleFactor for legends which is actualWidth / baseWidth where actualWidth is the width of the parent rectangle and baseWidth is the width of the SVG plot. If we set the width of the radar chart to 200px and the parent rectangle is 600px then the legend is going to be 300% its size. With the adaptive scaling we constrain the lower-upper scaling factor to 0.7 and 1.5 so it's never too small or too large.

dragonstyle commented on src/inspect_viz/view/beta/_scores_radar.py:
```diff
@@ -0,0 +1,297 @@
+
+    model_selection = Selection.single()
+
+    channels = {
```
I'm seeing a bunch of jankiness when using the hover to view channels (the popover is flickering and appearing and disappearing). This could be an issue with the way we've implemented the hover effect being confused by something about the marks. Could also be that the channel wants to be place only on 'dots' at the intersection of vertexes and lines?

alexandraabbas replied:
This is now solved by disabling tooltip on the other marks.

dragonstyle commented on src/inspect_viz/view/beta/_scores_radar.py:
```diff
@@ -0,0 +1,297 @@
+    if missing_columns:
+        raise ValueError(f"Required columns not found in data: {missing_columns}")
+
+    metrics = data.column_unique("metric")
```
I know we decided that passing metrics to the df function made the most sense, but now I am wondering whether it makes sense to support passing them as an arg here too (e.g. I could prepare a dataframe with all metrics, but then only choose to show some of the metrics)?

alexandraabbas replied:
In that case, this function would need to take a DataFrame as input instead of Data as we would need to filter the rows which we cannot do with Data. There's a specific schema we expect in scores_radar so I'd prefer users to rely on scores_radar_df to prepare their dataframe.

alexandraabbas committed:
Address comments

alexandraabbas committed:
Make scores_radar take Data only as input

alexandraabbas committed:
Set default plot size to 400px

alexandraabbas committed:
Add more detail to docs

Commit:
Simplify Legend Scaling

- Scale vs. SVG size rather than parent scale
- Never scale the legend up in size

dragonstyle commented:
1) I pushed a fix for the legend that I think will resolve the issues that the radar chart uncovered! 

2) I noticed that both Heatmap views and the sample tool views now have errors that I suspect may come from the changes to the `text` mark (though didn't fully track it down). To reproduce, just try rendering `docs/view-scores-heatmap.qmd`

Commit:
remove disused config

dragonstyle reviewed (changes requested):
Looks good - take a look at heatmap/samples tool use issue.

alexandraabbas committed:
Fix resolve_text_inputs failing to handle Transform objects

alexandraabbas committed:
Delete tests/inspect_viz/view/beta/test_scores_radar_data.py
